geode-core/src/distributedTest/java/org/apache/geode/cache30/CacheStatisticsPartitionedRegionDUnitTest.java

```diff
@@ -0,0 +1,577 @@
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 2 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(0.0f, rStats.getHitRatio(), 0.0f);
+
+    rStats.resetCounts();
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 0 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(0.0f, rStats.getHitRatio(), 0.0f);
+
+    region.put(key, value);
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 0 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(0, rStats.getMissCount());
+
+    region.get(key);
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 1 == rStats.getHitCount());
+    assertEquals(0, rStats.getMissCount());
+    assertEquals(1.0f, rStats.getHitRatio(), 0.0f);
+
+    region.get(key2);
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 1 == rStats.getMissCount());
+    assertEquals(1, rStats.getHitCount());
+    assertEquals(0.5f, rStats.getHitRatio(), 0.0f);
+
+    region.create(key3, null);
+    region.get(key3); // miss on existing entry
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 2 == rStats.getMissCount());
+    assertEquals(1, rStats.getHitCount());
+    assertEquals(2, rStats.getMissCount());
+    assertEquals(0.33f, rStats.getHitRatio(), 0.01f);
+
+    rStats.resetCounts();
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 0 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(0.0f, rStats.getHitRatio(), 0.0f);
+
+    region.invalidate(key);
+    region.get(key);
+
+    GeodeAwaitility.await().atMost(MSECS, TimeUnit.MILLISECONDS)
+        .until(() -> 1 == rStats.getMissCount());
+    assertEquals(0, rStats.getHitCount());
+    assertEquals(1, rStats.getMissCount());
+  }
+
+  /**
+   * Tests that the {@linkplain CacheStatistics#getLastAccessedTime last access time} and
+   * {@link CacheStatistics#getLastModifiedTime last modified time} are updated appropriately for a
+   * partitioned region with only one server.
+   */
+  @Test
+  public void testTimeStats() throws CacheException, InterruptedException {
+    final long ESTAT_RES = 100; // the resolution, in ms, of entry stats
+    String name = this.getUniqueName();
+    Object key = "KEY";
+    Object key2 = "KEY2";
+    Object value = "VALUE";
+    long before;
+    long after;
+    long oldBefore;
+    long oldAfter;
+
+    final int MSECS = 1000;
+
+    AttributesFactory factory = new AttributesFactory();
+    // factory.setScope(Scope.LOCAL);
+    factory.setStatisticsEnabled(true);
+
+    Region region = createPartitionedRegion(name, factory.create());
+    CacheStatistics rStats = region.getStatistics();
+
+    before = ((GemFireCacheImpl) getCache()).cacheTimeMillis();
```

Comment:
Before each setting of "before", I think it is worth waiting to see cacheTimeMillis return a different value. Basically we just want to make sure that each test records a new timestamp and that we are not just reading the timestamp set be the previous operation because the test runs so fast that all the ops get recorded at the same timestamp. I think you just have a simple named "waitForClockToChange" that reads it once and then loops until a subsequent call is not equal to the first call.

Reply:
Good point.